Added return raw sample functionality to qml.sample when observable is not provided.

Jaybsoni opened the pull request:
**Context:**

**Description of the Change:**

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**

josh146 commented:
Thanks @Jaybsoni for opening this PR 🙂 Let us know if you have any questions (feel free to comment directly on the code within the 'files changed' tab), or otherwise, once this PR is ready for a review.

> Fix bug with plotting the resultant circuit !

Out of curiosity, what is the status of this bug?

Jaybsoni commented:
Hi @josh146, 

Thanks for your patience, I got a bit busy and wasn't able to get as much work done as I would have liked these past few days! Here is an update for you: 

**Updates:** 
- Added unit tests for each of the files I have updated while implementing this feature 
    - Previously updated `measure.py` and `_qubit_device.py` so I added tests for each of these updates in `test_measure.py` and `test_qubit_device.py`  
- Fixed circuit visualization bug and added tests for the changes I made
    - Updated `representation_resolver.py` which can be found in the circuit_drawer module 
    - Added unit tests for these changes (in `test_representation_resolver.py`) 

**Here are some code snippets and context regarding the bug:**
The visualization bug was triggered when one would try to visualize a circuit which used the new `sample()` functionality
![circuit_drawer_visual bug](https://user-images.githubusercontent.com/21964864/124339093-cfdfab80-db79-11eb-9edd-9ff6ac8a44c0.png)

This bug was triggered because the object type passed into `representation_resolver.py` was a `MeasurementProcess` with `obs=None`. This would no longer trigger the initial if statement responsible for drawing the `sample` type. I added an additional sub-clause to account for this condition and represent it with the string representation: `raw`. Here is what the result looks like now: 
![circuit_drawer_visual update](https://user-images.githubusercontent.com/21964864/124339181-8f346200-db7a-11eb-95cd-0d22ec52f996.PNG)

**Here are some results from running the testing suite**:
At this point, I believe that the new feature has been fully integrated. In order to verify that it hasn't caused any other bugs in the library I ran the test suite. First, I ran the test suite on the master branch (which does not contain my updates) which is even with pennyLaneAI:master. I got the following results: 
![master_branch_test_suite](https://user-images.githubusercontent.com/21964864/124339243-30bbb380-db7b-11eb-943e-76c025159dc8.PNG)
![master_branch_test_suite2](https://user-images.githubusercontent.com/21964864/124339246-331e0d80-db7b-11eb-9e05-b76c5c84b4cd.PNG)

There seem to be some tests which fail or are skipped on the master branch itself. After looking through the error messages I have realized that most of these are due to me not having all of the optional dependencies (like dash["parallel"] or open fermion) installed in my environment. The numbers to keep track of are: ** 7 failed, 6656 passed, and 431 skipped ** 

Now here I run the test suite on the branch with my updated code: 
![raw_samples_branch_test_suite](https://user-images.githubusercontent.com/21964864/124339325-c8b99d00-db7b-11eb-867a-6d0913f683c1.PNG)
![raw_samples_branch_test_suite2](https://user-images.githubusercontent.com/21964864/124339326-cb1bf700-db7b-11eb-9956-02f1f3f77e97.PNG)

In this case we have ** 7 failed, 6665 passed, and 431 skipped **. The additional 9 tests which passed are the unit tests I added to test my updates. Since all of the error messages are identical to those resulting from running the test suite on the master branch, I am fairly confident that my feature implementation does not cause any other bugs. 

** To Do: **
- Update my local environment with optional dependencies and run the test suite again to make sure that there are no other errors! 
- Wait for any feedback from the Xanadu team on my implementation and updated code as required 

** Summary: ** 
I think the code is ready to be reviewed ! Hopefully I can integrate any changes you suggest and be ready to merge into master within the next few days. Feel free to provide feedback and let me know if you come across any other bugs (although I think we should be good as of now !). 

Thanks,

josh146 commented:
Awesome, great to hear @Jaybsoni! I'll mark this as ready for review, and someone from the team will be on hand to go through on Monday.

Also, great job tracking down and fixing the circuit drawing issue - the circuit drawing code is definitely quite complex!

josh146 reviewed (changes requested):
Hi @Jaybsoni, thanks for this great PR! I thought you tackled it with a compact and clean approach:

- I like how you re-used existing objects and functions where appropriate
- Nice attention to detail with both documentation, examples, and making sure the circuit drawer continues to work.

I have a couple of requested changes, but once addressed I think this is ready for merging:

- Don't forget to update the changelog, with (a) a description of the change, (b) an example codeblock, and (c) adding your name to the contributors section!

- It looks like you will also need to run black (`black -l 100 pennylane tests`) to ensure the formatting CI passes.

Jaybsoni commented on pennylane/_qubit_device.py:
```diff
@@ -791,6 +793,12 @@ def sample(self, observable, shot_range=None, bin_size=None):
+            samples = self._samples
+
+            if len(observable.wires) != 0:  # if wires are provided, then we only return samples from those wires
+                samples = self._samples[sample_slice, np.array(device_wires)]
```
Yes! We definitely need to account for the shot range, even if no wires were specified! 
Thanks for catching that bug!

Jaybsoni commented on pennylane/circuit_drawer/representation_resolver.py:
```diff
@@ -329,8 +329,11 @@ def operator_representation(self, op, wire):
+            if op.obs is not None:
+                op = op.obs
+            else:
+                return "raw"   # when no observable is provided we perform a raw measurement
```
For sure! I agree that basis makes more sense than 'raw', but it might still be ambiguous? (eg. which basis?) 
Another suggestion would be: 
'comp_basis'  -- for computational basis, although this might not be true for the CV devices 

I will commit 'basis' for now and maybe we can get some users or members of the team to comment on it?

Jaybsoni commented on pennylane/measure.py:
```diff
@@ -273,16 +273,21 @@ def circuit(x):
+def sample(op=None, wires=None):
     r"""Sample from the supplied observable, with the number of shots
     determined from the ``dev.shots`` attribute of the corresponding device.
+    If no observable is provided then sample from the device specific raw samples.
```
👍🏼

Jaybsoni commented on tests/circuit_drawer/test_representation_resolver.py:
```diff
@@ -203,6 +203,8 @@ def test_single_parameter_representation(self, unicode_representation_resolver,
             (qml.Toffoli(wires=[0, 2, 1]).inv(), 1, "X⁻¹"),
             (qml.Toffoli(wires=[0, 2, 1]).inv(), 0, "C"),
             (qml.Toffoli(wires=[0, 2, 1]).inv(), 2, "C"),
+            (qml.measure.sample(wires=[0, 1]), 0, "raw"),  # not providing an observable in
+            (qml.measure.sample(wires=[0, 1]), 1, "raw"),  # sample gets displayed as raw
```
Thanks ! 👍🏼

Jaybsoni commented on pennylane/measure.py:
```diff
@@ -300,17 +305,47 @@ def circuit(x):
+
     Args:
-        op (Observable): a quantum observable object
+        op (Observable or None): a quantum observable object
+        wires (Sequence[int] or int or None): the wires we wish to sample from
```
Sounds good, better to be redundant than to not provide the information sufficiently often.

Jaybsoni commented:
> Hi @Jaybsoni, thanks for this great PR! I thought you tackled it with a compact and clean approach:
> 
> * I like how you re-used existing objects and functions where appropriate
> * Nice attention to detail with both documentation, examples, and making sure the circuit drawer continues to work.
> 
> I have a couple of requested changes, but once addressed I think this is ready for merging:
> 
> * Don't forget to update the changelog, with (a) a description of the change, (b) an example codeblock, and (c) adding your name to the contributors section!
> * It looks like you will also need to run black (`black -l 100 pennylane tests`) to ensure the formatting CI passes.

Hi @josh146, 

I incorporated all of the changes you suggested, double checked that the test suite runs without errors, re-formatted the files using `black` and added information about the change to the ChangeLog. Thanks for all of your comments, and your help throughout this project! I had a lot of fun diving into Pennylane's source code! Let me know if there are any other finishing touches that I need to put on the project. Otherwise I think we are good to merge ! 

Thanks,

Jaybsoni commented:
On a side note: It seems that CodeFactor keeps failing, but it seems to be failing on some functions that I haven't changed. Not sure what is wrong here ...

albi3ro reviewed (changes requested):
This looks excellent 👍  and I'm very happy to get this change in.

Some requests for polishing it up:

For the `circuit_1` example, the drawing does not work as expected.

`Sample[basis]` shows up when you explicitly set the wires, but it will not show up when you do not explicitly set the wires.

Also, you can call

```
@qml.qnode(dev)
def circ():
    return qml.sample(wires=[])
```

This will return samples for all wires.  Intuitively, I would expect that to return samples for no wires.

Similarly, if you do not specify wires, `circuit_1.qtape.operations[0]` will be `sample(wires=[])`.  This implies "no wires" instead of "all wires" to me.

josh146 commented on .github/CHANGELOG.md:
```diff
@@ -26,6 +26,48 @@
 * Ising YY gate functionality added.
   [(#1358)](https://github.com/PennyLaneAI/pennylane/pull/1358)
+  
+* Added functionality to `qml.sample()` to extract samples from the basis states of
+the device (currently only for qubit devices). Additionally, `wires` can be
+specified to only return samples from those wires. 
+[(#1441)](https://github.com/PennyLaneAI/pennylane/pull/1441)
```
```suggestion
* Added functionality to `qml.sample()` to extract samples from the basis states of
  the device (currently only for qubit devices). Additionally, `wires` can be
  specified to only return samples from those wires. 
  [(#1441)](https://github.com/PennyLaneAI/pennylane/pull/1441)
```

When writing lists in markdown, the continuation lines should match the indentation of the first line!

josh146 commented on .github/CHANGELOG.md:
````diff
@@ -26,6 +26,48 @@
+specified to only return samples from those wires. 
+[(#1441)](https://github.com/PennyLaneAI/pennylane/pull/1441)
+
+```python
+dev = qml.device("default.qubit", wires=3, shots=5)
+
+@qml.qnode(dev)
+def circuit_1():
+    qml.Hadamard(wires=0)
+    qml.Hadamard(wires=1)
+    return qml.sample()    # no observable provided
+
+@qml.qnode(dev)
+def circuit_2():
+    qml.Hadamard(wires=0)
+    qml.Hadamard(wires=1)
+    return qml.sample(wires=[0,2])    # no observable provided and wires specified 
+``` 
````
Same here - this entire codeblock (including the backticks) needs to be indented by two spaces

josh146 commented on .github/CHANGELOG.md:
```diff
@@ -26,6 +26,48 @@
+def circuit_1():
+    qml.Hadamard(wires=0)
+    qml.Hadamard(wires=1)
+    return qml.sample()    # no observable provided
```
```suggestion
    return qml.sample()
```

I would say no need to specify this here :)

josh146 commented on .github/CHANGELOG.md:
```diff
@@ -26,6 +26,48 @@
+   [0, 0]]
+
+  >>> print(qml.draw(circuit_2))
+ 0: ──H──╭┤ Sample[basis]
+ 1: ──H──│┤
+ 2: ──H──╰┤ Sample[basis]
```
```suggestion
   0: ──H──╭┤ Sample[basis]
   1: ──H──│┤
   2: ──H──╰┤ Sample[basis]
```

josh146 commented on tests/test_qubit_device.py:
```diff
@@ -577,6 +577,34 @@ def test_correct_custom_eigenvalues(
         assert np.array_equal(res, np.array([-1, 1]))
 
+    def test_sample_with_no_observable_and_no_wires(
+        self, mock_qubit_device_with_original_statistics, monkeypatch, tol
```
```suggestion
        self, mock_qubit_device_with_original_statistics, tol
```

josh146 commented on tests/test_qubit_device.py:
```diff
@@ -577,6 +577,34 @@ def test_correct_custom_eigenvalues(
+        assert np.array_equal(res, generated_samples)
+
+    def test_sample_with_no_observable_and_with_wires(
+        self, mock_qubit_device_with_original_statistics, monkeypatch, tol
```
```suggestion
        self, mock_qubit_device_with_original_statistics, tol
```

josh146 reviewed (approved):
Thanks for implementing the suggested changes @Jaybsoni! This looks good to go from my end now, pending a few small changes :)

Jaybsoni commented:
Hi @albi3ro, 
Thanks for the comments! 

> For the `circuit_1` example, the drawing does not work as expected.
> 
> `Sample[basis]` shows up when you explicitly set the wires, but it will not show up when you do not explicitly set the wires.

Nice job catching that display bug! I will look into it asap.

For your second comment:
> @qml.qnode(dev)
> def circ():
>     return qml.sample(wires=[])
> ```
> 
> This will return samples for all wires. Intuitively, I would expect that to return samples for no wires.

This might be a little bit complicated. Currently, `MeasurementProcess` instances can't have a None type for the wires. This can be seen in the code (line 54 of measure.py) as: `self._wires = wires or Wires([])`. Thus providing `wires = []` and `wires = None` is functionally identical in `qml.sample()`. I am a little bit hesitant to make a changes to this class since it is so fundamental and I haven't had a chance to explore the entire code base. That being said, I can offer two options: 


- Since sampling with `wires=[]` doesn't make physical sense, it might be more appropriate to just throw an error in `qml.sample` which tells the user that if they don't specify the observable then the wires must be non-empty or None. 

- I can try and augment `MeasurementProcess` such that it allows for None type wires, and then use that information to return None when `qml.sample(wires=[])` is called. I just have to make sure we won't throw any errors in any other functions which will have access to our object. 

Let me know which one makes more sense to implement in this case. If you have other suggestions I would love to hear them! 
Thanks,

albi3ro commented:
@Jaybsoni  Thanks for looking into that.  Given the trouble of fixing the `sample(wires=[)` problem, I think we can leave that be.  

Therefore, only the drawing problem needs to be fixed.

josh146 commented:
> Given the trouble of fixing the sample(wires=[) problem, I think we can leave that be.

Yep, I think this makes sense in the meantime. This way we can come back and fix/update the `MeasurementProcess` class in a separate PR :)

antalszava commented:
Hi @Jaybsoni, cool addition! :slightly_smiling_face: :tada:  I got wondering if drawing `Sample[basis]` at the end of the circuit for having `qml.sample()` could somewhat lead to ambiguity regarding which basis it would be? Maybe it's me, just thought of leaving a comment up for consideration, wouldn't seem to be a blocker.

Jaybsoni commented:
Hi all, 

Sorry for the delay, I got busy with work. 

This bug was a tricky one to pin down, especially since the plotting modules are very inter-connected. The root cause was in `circuit_graph.py` (lines 137 - 159), when initializing the attribute `self._grid`. This attribute represents the quantum circuit. Prior to my change, we would look through each operator or observable provided, iterate over the wires it acts upon and then add them accordingly to `self._grid`. This would fail when we call `qml.sample()` because implicitly `wires = qml.wires.Wires([])`. I added a simple catch to identify when a `qml.sample()` was called and no wires were specified. In this case the operation is replaced with a sample operation where all available wires are provided instead.  

Here is a snapshot of the result: 
![circuit_drawer_visual update group print](https://user-images.githubusercontent.com/21964864/126026567-de29b91a-1401-4d30-883a-c168e5b49211.PNG)

@albi3ro, I think everything should be good to go as of now! Let me know if there is anything I should change / fix about my implementation! 

Thanks,

albi3ro reviewed (approved):
Thanks for figuring that out :) I know the circuit drawer stuff is a bit of tricky mess right now.

Approve :100:

josh146 commented on pennylane/circuit_graph.py:
```diff
@@ -145,11 +146,17 @@ def __init__(self, ops, obs, wires, par_info=None, trainable_params=None):
+
+                elif op.return_type is qml.operation.Sample and op.wires == Wires([]):
+                    # Sampling without specifying wires is treated as sampling all wires
+                    op = qml.measure.MeasurementProcess(qml.operation.Sample, wires=wires)
```
Thanks @Jaybsoni for all your work so far, this is very close to being merged in!

However, this line looks like it is not being tested, which is blocking the CI checks

Jaybsoni commented on tests/circuit_graph/test_circuit_graph.py:
```diff
@@ -93,6 +93,40 @@ def circuit_measure_multiple_with_max_twice():
     )
 
 
+def measurement_process_equality(mp1, mp2):
+    """Helper function to check for equality between two MeasurementProcess objects."""
+    same_return_type = mp1.return_type == mp2.return_type
+    same_obs = mp1.obs == mp2.obs
+    same_wires = mp1._wires == mp2._wires
+    same_eigs = mp1._eigvals == mp2._eigvals
+    same_name = mp1.name == mp2.name
+    same_data = mp1.data == mp2.data
+
+    if same_return_type and same_obs and same_wires and same_eigs and same_name and same_data:
+        return True
+
+    return False
+
+
+def lst_mp_equality(lst1, lst2):
+    """Wrapper function to check if two lists are identical when the
+    lists contain MeasurementProcess objects."""
+    assert len(lst1) == len(lst2)
+    lsts_same = True
+
+    for index in range(len(lst1)):
+        if not isinstance(lst1[index], type(lst2[index])):  # if they are different types
+            return False  # then return false
+
+        elif isinstance(lst1[index], qml.measure.MeasurementProcess):
+            lsts_same = lsts_same and measurement_process_equality(lst1[index], lst2[index])
+
+        else:
+            lsts_same = lsts_same and (lst1[index] == lst2[index])
+
+    return lsts_same
```
I added a test to ensure that the line of code I added would actually generate a circuit graph object such that the object generated by `qml.sample()` is identical to the one generated when `qml.sample(wires=[0,1,2])` with all the wires explicitly mentioned. 

When adding this implementation I came across a little problem: 
`obj1 = qml.measure.sample()` 
`obj2 = qml.measure.sample()`
`obj1 == obj2  #returns false` 

This is probably because the equality operator is checking if they are references to the same object as opposed to comparing their fields. For this reason I had to implement two other helper functions in the text file to allow me to more easily compare these MeasurementProcess objects without compromising code readability. 

I tried over-writing the `==` operator in the MeasurementProcess class, but this lead to other errors in other files, so I decide to go with this solution. Let me know what you think! 

Thanks,

josh146 replied:
Thanks @Jaybsoni! I imagine defining `__eq__` for the `MeasurementProcess` class led to some issues with hashing of the object within the codebase 🤔 Interesting.

For now, this is a decent solution!

josh146 reviewed (approved):
Thanks for this great contribution @Jaybsoni!
